### ISSUE TITLE
StyleOptimizedVectorTileFactory how to configure?

### ISSUE THREAD

**Issue by AliasSCM** [mapbox/mapbox-unity-sdk#282]
Ive updated to the new version of the mapbox sdk and im stuck at trying to configure the StyleOptimizedVectorTileFactory. Seems like the tutorials are still referring to the Meshfactory.

Specificaly, im trying to get the StyleId , modifiedData and name to configure the StyleOptimizedVectorTileFactory but im unable to get the details on the same for my styles in my mapbox account. I tried the search function as well but it just keeps on "searching" without showing any results. Ive attached a screenshot here:

![screen shot 2017-10-10 at 7 46 28 pm](https://user-images.githubusercontent.com/15645920/31391453-32d8c40e-adf4-11e7-877b-7212feb66252.png)

Im using Unity 2017.2.0b11 and mapbox-unity-sdk version 1.2.0...

**Comment by david-rhodes** [comment]
@AliasSCM I see a console error at the bottom of the Unity editor. It would be helpful if you posted the entire call stack here. This error is the reason your search did not yield any results.

A few other questions that could help identify the issue:
1. Do you perhaps only have one style in your mapbox account?
2. If you close Unity and try again, do you get the same error?

In the meantime, you can download the `style.json` from Mapbox studio and manually enter the required information (Id, Name, Modified).

**Comment by AliasSCM** [comment]
@david-rhodes Here is the full log from the console

JsonSerializationException: Cannot deserialize the current JSON object (e.g. {"name":"value"}) into type 'System.Object[]' because the type requires a JSON array (e.g. [1,2,3]) to deserialize correctly.
To fix this error either change the JSON to a JSON array (e.g. [1,2,3]) or change the deserialized type so that it is a normal .NET type (e.g. not a primitive type like integer, not a collection type like an array or List<T>) that can be deserialized from a JSON object. JsonObjectAttribute can also be added to the type to force it to deserialize from a JSON object.
Path 'message', line 1, position 11.
Mapbox.Json.Serialization.JsonSerializerInternalReader.CreateObject (Mapbox.Json.JsonReader reader, System.Type objectType, Mapbox.Json.Serialization.JsonContract contract, Mapbox.Json.Serialization.JsonProperty member, Mapbox.Json.Serialization.JsonContainerContract containerContract, Mapbox.Json.Serialization.JsonProperty containerMember, System.Object existingValue)
Mapbox.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal (Mapbox.Json.JsonReader reader, System.Type objectType, Mapbox.Json.Serialization.JsonContract contract, Mapbox.Json.Serialization.JsonProperty member, Mapbox.Json.Serialization.JsonContainerContract containerContract, Mapbox.Json.Serialization.JsonProperty containerMember, System.Object existingValue)
Mapbox.Json.Serialization.JsonSerializerInternalReader.Deserialize (Mapbox.Json.JsonReader reader, System.Type objectType, Boolean checkAdditionalContent)

Do you perhaps only have one style in your mapbox account?
No i have 5 custom styles. 

If you close Unity and try again, do you get the same error?
Yes, i have tried this multiple times but still i get the same error


I was able to get the details from the style.json file, but at this point im not sure how to configure the StyleOptimizedVectorTileFactory correctly. When i run the scene i just get base raster map and not the building overlays that i want. The tutorials im following still uses the deprecated MeshFactory..

This is how my StyleOptimizedVectorTileFactory is setup currently...

![screen shot 2017-10-11 at 11 51 59 am](https://user-images.githubusercontent.com/15645920/31424963-b03d1b58-ae7a-11e7-9fb1-5b4cbd0a086f.png)

**Comment by BergWerkGIS** [comment]
@AliasSCM wrt searching for a style:
Seeing you are on a beta version, I tried with **`Unity@2017.3.0b4`** and `SDK@1.2.0` and don't get any errors.

Is trying to upgrade Unity an option for you?

![image](https://user-images.githubusercontent.com/4549888/31435956-9ae06d4e-ae81-11e7-85ed-fbe18850104f.png)

**Comment by AliasSCM** [comment]
@BergWerkGIS isnt Unity@2013.3.04b in beta still?

Updating is not that big of an issue for me, but i still am slightly skeptical that other functionalities will get affected in my project. I will give it a shot anyway...

Also, i am trying to drop the VectorTileFactoryStyleOptimized (one last effort before i upgrade!) and just use the VectorTileFactory but i still dont see the buildings i want. I dont see any error in this case...

![screen shot 2017-10-11 at 6 33 04 pm](https://user-images.githubusercontent.com/15645920/31442151-1259d6e0-aeb3-11e7-9f56-d0f1f6299aec.png)
![screen shot 2017-10-11 at 6 33 10 pm](https://user-images.githubusercontent.com/15645920/31442155-14686b54-aeb3-11e7-8e48-6dc515c92d29.png)
![screen shot 2017-10-11 at 6 33 19 pm](https://user-images.githubusercontent.com/15645920/31442158-161bd184-aeb3-11e7-9e36-c64982a66712.png)

**Comment by BergWerkGIS** [comment]
> @BergWerkGIS isnt Unity@2013.3.04b in beta still?

Yeah, I tried the lastest beta as you are on a beta yourself:
![image](https://user-images.githubusercontent.com/4549888/31443221-1f86506a-ae99-11e7-8069-471bf72835a7.png)

Latest stable as of today according to https://store.unity.com/download is `2017.1.1`

Also we had another report about broken SDK with `v2017.2 beta` (https://github.com/mapbox/mapbox-unity-sdk/issues/233) that was solved by upgrading to the lastest beta.
So that might definitely be worth a short.

-----

@david-rhodes will answer about the factories.

**Comment by david-rhodes** [comment]
@AliasSCM I don't see a screenshot of your `Map` component or the `MapVisualizer` asset you are using. If you don't see buildings generated with a basic `VectorTileFactory` (which is just a renamed `MeshFactory`), then the fault probably lies in the assigned factories, or some layer may be disabled. Please check that your factory is assigned in the `MapVisualizer`.

**Comment by david-rhodes** [comment]
Perhaps you could email me your raw JSON? You can log this data here: https://github.com/mapbox/mapbox-unity-sdk/blob/develop/sdkproject/Assets/Mapbox/Unity/Editor/StyleSearchWindow.cs#L167. 

Send it to david.rhodes@mapbox.com. I'll try to see why the JSON parsing is failing. Thanks!

**Comment by david-rhodes** [comment]
A screenshot of your map hierarchy from the `Map Editor` would also be helpful. Thanks!

**Comment by AliasSCM** [comment]
@BergWerkGIS Updated to Unity@2017.3.0b2 and the results are the same...

![screen shot 2017-10-11 at 11 15 44 pm](https://user-images.githubusercontent.com/15645920/31457662-090f8472-aedb-11e7-99d9-afaa0b696fb8.png)

I tried using the VectorTileFactory as well but could not see the buildings, just the base map with the terrain.

**Comment by AliasSCM** [comment]
@david-rhodes ive mailed the log to the email you requested.

Here are the screenshots of my MapVisualizer , Map Object and the Merged Modifier Stack...

![screen shot 2017-10-12 at 9 50 51 am](https://user-images.githubusercontent.com/15645920/31479224-5edaad7a-af33-11e7-9e4e-ae8e9f98d991.png)

![screen shot 2017-10-12 at 9 51 10 am](https://user-images.githubusercontent.com/15645920/31479227-618094e0-af33-11e7-8dc1-6012fb4fb478.png)

![screen shot 2017-10-12 at 9 51 38 am](https://user-images.githubusercontent.com/15645920/31479229-636a3c16-af33-11e7-8ae6-76922caf47be.png)

**Comment by AliasSCM** [comment]
I was able to resolve this issue. The fix was very simple. For some reason, one of my tokens had got revoked from my mapbox account. When i changed the mapbox configuration to my "Default Public Token" and "Cleared Cache" everything worked as expected. 

I did some debugging to get to this conclusion (seems so obvious now!), and while debugging this issue i had some findings, not sure how useful it will be to you guys, but anyway...

The reason i wasnt seeing any buildings was that the response.data (in Tile.cs -> HandleTileResponse )was coming in as null. Calling this url (https://api.mapbox.com/v4/mapbox.terrain-rgb/16/10485/25327.pngraw?events=true) verified this for me.

This obviously happened because my token had expired , and Im assuming that the map images and terrain data were cached somewhere so that used to come up nicely. As soon as i cleared my cache everything went pink. It would help if there was a console error saying that the token has expired...

But so far, loving the Unity SDK you guys have put together! I have been waiting for this for a long time :)
